session/pingpong/invoice_tracker.go

```diff
@@ -213,43 +292,54 @@ func (it *InvoiceTracker) Start() error {
 		return ErrConsumerNotRegistered
 	}
 
-	fees, err := it.feeProvider.FetchSettleFees()
+	fees, err := it.deps.FeeProvider.FetchSettleFees()
 	if err != nil {
 		return errors.Wrap(err, "could not fetch fees")
 	}
 	it.transactorFee = fees.Fee
 
-	fee, err := it.bcHelper.GetAccountantFee(common.HexToAddress(it.accountantID.Address))
+	fee, err := it.deps.BlockchainHelper.GetAccountantFee(common.HexToAddress(it.deps.AccountantID.Address))
 	if err != nil {
 		return errors.Wrap(err, "could not get accountants fee")
 	}
 
-	if fee > it.maxAllowedAccountantFee {
-		log.Error().Msgf("Accountant fee too large, asking for %v where %v is the limit", fee, it.maxAllowedAccountantFee)
+	if fee > it.deps.MaxAllowedAccountantFee {
+		log.Error().Msgf("Accountant fee too large, asking for %v where %v is the limit", fee, it.deps.MaxAllowedAccountantFee)
 		return ErrAccountantFeeTooLarge
 	}
 
-	err = it.generateInitialInvoice()
+	agreementID, err := it.deps.InvoiceStorage.GetNewAgreementID(it.deps.ProviderID)
 	if err != nil {
-		return errors.Wrap(err, "could not generate initial invoice")
+		return errors.Wrap(err, "could not get new agreement id")
 	}
 
+	it.agreementID = agreementID
+
+	emErrors := make(chan error)
+	go func() {
+		emErrors <- it.listenForExchangeMessages()
+	}()
+
 	// give the consumer a second to start up his payments before sending the first request
 	firstSend := time.After(time.Second)
 	for {
 		select {
 		case <-firstSend:
-			err := it.sendInvoiceExpectExchangeMessage()
+			err := it.sendInvoice()
 			if err != nil {
 				return err
 			}
 		case <-it.stop:
 			return nil
-		case <-time.After(it.chargePeriod):
-			err := it.sendInvoiceExpectExchangeMessage()
+		case <-time.After(it.deps.ChargePeriod):
+			err := it.sendInvoice()
 			if err != nil {
 				return err
 			}
+		case emErr := <-emErrors:
+			if emErr != nil {
```

Comment:
is this condition needed?

Reply:
yes, because on channel close we'll get a nil value